be/src/exprs/dict_query_expr.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exprs/dict_query_expr.h"
+
+#include "column/chunk.h"
+#include "column/column.h"
+#include "column/column_helper.h"
+#include "column/column_viewer.h"
+#include "gutil/casts.h"
+#include "storage/chunk_helper.h"
+#include "storage/table_reader.h"
+#include "udf/java/utils.h"
+
+namespace starrocks {
+
+DictQueryExpr::DictQueryExpr(const TExprNode& node) : Expr(node), _dict_query_expr(node.dict_query_expr) {}
+
+DictQueryExpr::~DictQueryExpr() {}
+
+StatusOr<ColumnPtr> DictQueryExpr::evaluate_checked(ExprContext* context, Chunk* ptr) {
+    Columns columns(children().size());
+    size_t size = ptr != nullptr ? ptr->num_rows() : 1;
+    for (int i = 0; i < _children.size(); ++i) {
+        columns[i] = _children[i]->evaluate(context, ptr);
+    }
+
+    ColumnPtr res;
+    for (auto& column : columns) {
+        if (column->is_constant()) {
+            column = ColumnHelper::unpack_and_duplicate_const_column(size, column);
+        }
+    }
+    ChunkPtr key_chunk = ChunkHelper::new_chunk(_key_slots, size);
+    key_chunk->reset();
+    for (int i = 0; i < _dict_query_expr.key_fields.size(); ++i) {
+        ColumnPtr key_column = columns[1 + i];
+        key_chunk->update_column_by_index(key_column, i);
+    }
+
+    for (auto& column : key_chunk->columns()) {
+        if (column->is_nullable()) {
+            column = ColumnHelper::update_column_nullable(false, column, column->size());
+        }
+    }
+
+    std::vector<bool> found;
+    ChunkPtr value_chunk = ChunkHelper::new_chunk(_value_schema, key_chunk->num_rows());
+
+    Status status = _table_reader->multi_get(*key_chunk, {_dict_query_expr.value_field}, found, *value_chunk);
+    if (!status.ok()) {
+        // todo retry
+        LOG(WARNING) << "fail to execute multi get: " << status.detailed_message();
+        return status;
+    }
+    res = _value_schema.field(0)->with_nullable(true)->create_column();
+
+    int res_idx = 0;
+    for (int idx = 0; idx < size; ++idx) {
+        if (found[idx]) {
+            res->append_datum(value_chunk->get_column_by_index(0)->get(res_idx));
+            res_idx++;
+        } else {
+            res->append_nulls(1);
+        }
+    }
+
+    return res;
+}
+
+Status DictQueryExpr::prepare(RuntimeState* state, ExprContext* context) {
```

Comment:
Can we get the routing information in DictQueryExpr::prepare through a RPC call instead of get it from TDictQueryExpr? In this way, we need not refactor the thrift code.

Reply:
modified